### PR TITLE
feature - v3 tranche contract

### PR DESCRIPTION
**Pull request by ilinzweilin**
to be done:
- more tests covering edge cases 
- add investor auth to supplyOrder & redeemOrder & disburse to support investor memberlist

### CODE REVIEW

File: src/lender/tranche/v3/ticker.sol [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    uint public epochCount;
+
+    constructor() public {
+        // 00:00 next day first epoch starts
+        firstEpochTimestamp = normalizeTimestamp(now);
+    }
+
+    // normalizes timestamp to 00:00
+    function normalizeTimestamp(uint timestamp) internal returns (uint) {
+        return safeAdd(rmul((1 days), rdiv(timestamp, 1 days)), 1 days);
```
Comment (manuelpolzhofer) [inline src/lender/tranche/v3/ticker.sol:31]:
```suggestion
        return rmul((1 days), rdiv(timestamp, 1 days));
```
The last plus 1 day is not required. 

I had it initially in the spec to sort of start the first epoch at 00:00 but for normalize its not required i think.

File: src/lender/tranche/v3/ticker.sol [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    }
+    
+    function currentEpoch() public returns (uint) {
```
Comment (manuelpolzhofer) [inline src/lender/tranche/v3/ticker.sol:34]: lets add some unit tests for this

File: src/lender/tranche/v3/tranche.sol [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+contract TickerLike {
+    function currentEpoch() public returns (uint);
+}
+
+contract Tranche is Math, Auth {
+mapping (uint => Epoch) public epochs;
+struct Epoch {
+    uint totalRedeem;
+    uint totalSupply;
+    // denominated in WAD
```
Comment (manuelpolzhofer) [inline src/lender/tranche/v3/tranche.sol:38]:
```suggestion
    // denominated in RAY
```

File: src/lender/tranche/v3/tranche.sol [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    // percentage ONE == 100%
+    uint redeemFulfillment;
+    // denominated in WAD
```
Comment (manuelpolzhofer) [inline src/lender/tranche/v3/tranche.sol:41]:
```suggestion
    // denominated in RAY
```

File: src/lender/tranche/v3/tranche.sol [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    wards[msg.sender] = 1;
+    token = ERC20Like(token_);
+    currency = ERC20Like(currency_);
+    ticker = TickerLike(ticker_);
+    self = address(this);
+}
+
+function supplyCurrencyAmount(uint epochID, address addr) public view returns (uint) {
+    Epoch storage epoch = epochs[epochID];
+    return epoch.supplyCurrencyAmount[addr];
```
Comment (manuelpolzhofer) [inline src/lender/tranche/v3/tranche.sol:65]:
```suggestion
    return epochs[epochID].supplyCurrencyAmount[addr];
```

File: src/lender/tranche/v3/tranche.sol [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+}
+
+function redeemTokenAmount(uint epochID, address addr) public view returns (uint) {
+    Epoch storage epoch = epochs[epochID];
+    return epoch.redeemTokenAmount[addr];
```
Comment (manuelpolzhofer) [inline src/lender/tranche/v3/tranche.sol:70]:
```suggestion
    return epochs[epochID].redeemTokenAmount[addr];
```

File: src/lender/tranche/v3/tranche.sol [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+function depend(bytes32 contractName, address addr) public auth {
+    if (contractName == "token") { token = ERC20Like(addr); }
+    if (contractName == "currency") { currency = ERC20Like(addr); }
+    if (contractName == "ticker") { ticker = TickerLike(addr); }
+    else revert();
+}
+
+// supplyOrder function can be used to place or revoke an supply 
+function supplyOrder(uint epochID, uint supplyAmount) public {
```
Comment (manuelpolzhofer) [inline src/lender/tranche/v3/tranche.sol:89]:
```suggestion
function supplyOrder(uint epochID, uint newSupplyAmount) public {
```

File: src/lender/tranche/v3/tranche.sol [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+    require((epochID >= ticker.currentEpoch()), "epoch-already-over");
+    uint currentSupplyAmount = epochs[epochID].supplyCurrencyAmount[msg.sender];
+    epochs[epochID].supplyCurrencyAmount[msg.sender] = supplyAmount;
+    epochs[epochID].totalSupply = safeAdd(safeSub(epochs[epochID].totalSupply, currentSupplyAmount), supplyAmount);
```
Comment (manuelpolzhofer) [inline src/lender/tranche/v3/tranche.sol:93]: I think you need to update the `totalSupply` with the delta instead of the entire amount

File: src/lender/tranche/v3/tranche.sol [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+    if (supplyAmount > currentSupplyAmount) {
+        uint delta = safeSub(supplyAmount, currentSupplyAmount);
+        require(currency.transferFrom(msg.sender, self, delta), "currency-transfer-failed");
```
Comment (manuelpolzhofer) [inline src/lender/tranche/v3/tranche.sol:96]: why no `delta>0` check here?

File: src/lender/tranche/v3/tranche.sol [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+    }
+}
+
+// redeemOrder function can be used to place or revoke a redeem
+function redeemOrder(uint epochID, uint redeemAmount) public {
+    require((epochID >= ticker.currentEpoch()), "epoch-already-over");
+    
+    uint currentRedeemAmount = epochs[epochID].redeemTokenAmount[msg.sender];
+    epochs[epochID].redeemTokenAmount[msg.sender] = redeemAmount;
+    epochs[epochID].totalRedeem = safeAdd(safeSub(epochs[epochID].totalRedeem, currentRedeemAmount), redeemAmount);
```
Comment (manuelpolzhofer) [inline src/lender/tranche/v3/tranche.sol:111]: same here for `totalRedeem`

File: src/lender/tranche/v3/tranche.sol [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+    
+     if (redeemAmount > currentRedeemAmount) {
+        uint delta = safeSub(redeemAmount, currentRedeemAmount);
+        require(token.transferFrom(msg.sender, self, delta), "token-transfer-failed");
+        token.burn(self, redeemAmount);
```
Comment (manuelpolzhofer) [inline src/lender/tranche/v3/tranche.sol:116]: We should burn and mint only in the epoch execute because it impacts the tokenPrice.

Assume Alice and Bob both invest 50 DAI for 1.00DAI token Price. Pool would be 100 DAI for 100 Token.

If Alice would now make an redeemOrder we would burn the 50 Token and the token price would be 100/50 = 2.0.

Alice would receive 100 DAI for her 50 token.

File: src/lender/tranche/v3/tranche.sol [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+    uint tokenAmount = calcTokenDisbursement(epochID);
+    
+    epochs[epochID].supplyCurrencyAmount[msg.sender] = 0;
+    if (currencyAmount > 0) {
+        require(currency.transferFrom(self, msg.sender, currencyAmount), "currency-transfer-failed"); 
+    }
+    
+    epochs[epochID].redeemTokenAmount[msg.sender] = 0;
+    if (tokenAmount > 0) {
+        token.mint(msg.sender, tokenAmount);
```
Comment (manuelpolzhofer) [inline src/lender/tranche/v3/tranche.sol:142]: here the same. I think we will get in trouble with the tokenPrice, if we do it like that.

File: src/lender/tranche/v3/tranche.sol (hunk 8 above)
Comment (ilinzweilin) [reply src/lender/tranche/v3/tranche.sol:91]: haha we are doing it: substracting old amount and adding new amount

File: src/lender/tranche/v3/tranche.sol (hunk 9 above)
Comment (ilinzweilin) [reply src/lender/tranche/v3/tranche.sol:94]: Because supplyAmount is bigger then currentSupplyAmount so delta <= 0 is not possible

File: src/lender/tranche/v3/tranche.sol [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+    uint currentSupplyAmount = epochs[epochID].supplyCurrencyAmount[msg.sender];
+    epochs[epochID].supplyCurrencyAmount[msg.sender] = supplyAmount;
+    epochs[epochID].totalSupply = safeAdd(safeSub(epochs[epochID].totalSupply, currentSupplyAmount), supplyAmount);
+    if (supplyAmount > currentSupplyAmount) {
+        uint delta = safeSub(supplyAmount, currentSupplyAmount);
+        require(currency.transferFrom(msg.sender, self, delta), "currency-transfer-failed");
+        return;
+    } 
+    uint delta = safeSub(currentSupplyAmount, supplyAmount);
+    if (delta > 0) {
```
Comment (ilinzweilin) [inline src/lender/tranche/v3/tranche.sol:98]: here delta = 0 possible, since currentSupplyAmoun can be equal to supplyAmount

File: src/lender/tranche/v3/tranche.sol [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+        token.burn(self, diff);
+        return;
+    }
+
+    // mint tokens that are required for disbursement
+    uint diff = safeSub(mintAmount, burnAmount);
+    if (diff > 0) {
+        token.mint(self, diff);
+    }
+}
```
Comment (manuelpolzhofer) [inline src/lender/tranche/v3/tranche.sol:178]: What is currently still missing is the currency take from reserve or give to reserve depending on the currency delta

Comment (ilinzweilin) [reply src/lender/tranche/v3/tranche.sol:178]: true!

File: src/lender/tranche/v3/tranche.sol [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+    // burn tokens that are not needed for disbursement
+    if (burnAmount > mintAmount) {
+        uint diff = safeSub(burnAmount, mintAmount);
+        token.burn(self, diff);
+        return;
+    }
+    // mint tokens that are required for disbursement
+    uint diff = safeSub(mintAmount, burnAmount);
+    if (diff > 0) {
+        token.mint(self, diff);
```
Comment (manuelpolzhofer) [inline src/lender/tranche/v3/tranche.sol:184]: I think you could make one internal helper function which returns the diff and a flag would you should do with it.

Because the code is very similar

Comment (manuelpolzhofer) [reply src/lender/tranche/v3/tranche.sol:184]: maybe we could violate our rule here and just return an `int` depending if it is positive or negative we can proceed

File: src/lender/tranche/v3/tranche.sol [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+    }
+}
+
+// adjust currency balance after epoch execution -> receive/send currency from/to reserve
+function adjustCurrencyBalance(uint epochID, uint supplyFulfillment_, uint redeemFulfillment_, uint tokenPrice_) internal {
```
Comment (manuelpolzhofer) [inline src/lender/tranche/v3/tranche.sol:192]:
```suggestion
function adjustCurrencyBalance(uint epochID) internal {
```
Either we pass the parameters or we access them from the state.

File: src/lender/tranche/v3/tranche.sol [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+    epochs[epochID].supplyFulfillment = supplyFulfillment_;
+    epochs[epochID].redeemFulfillment = redeemFulfillment_;
+    epochs[epochID].tokenPrice = tokenPrice_;
+     
+    adjustTokenBalance(epochID, supplyFulfillment_, redeemFulfillment_, tokenPrice_);
+    adjustCurrencyBalance(epochID, supplyFulfillment_, redeemFulfillment_, tokenPrice_);
+}
+
+// adjust token balance after epoch execution -> min/burn tokens
+function adjustTokenBalance(uint epochID, uint supplyFulfillment_, uint redeemFulfillment_, uint tokenPrice_) internal {
```
Comment (manuelpolzhofer) [inline src/lender/tranche/v3/tranche.sol:173]:
```suggestion
function adjustTokenBalance(uint epochID) internal {
```
